### PR TITLE
Add bow and crossbow hooks for custom items

### PR DESCRIPTION
**Pull request by OroArmor**
This pull request is to make the creation for bows and crossbows way easier. Just implementing the interface will allow any bow/crossbow to easily edit the hard-coded features from Minecraft.

### CODE REVIEW

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/AbstractSkeletonEntityMixin.java [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+import net.minecraft.item.BowItem;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.Items;
+
+@Mixin(AbstractSkeletonEntity.class)
+public class AbstractSkeletonEntityMixin {
+
+	@Redirect(method = "updateAttackType", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;getItem()Lnet/minecraft/item/Item;"))
+	private Item renderItem(ItemStack itemStack) {
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/AbstractSkeletonEntityMixin.java]: `renderItem`?

Like flip the tracked data flag to show a skeleton is pulling the arrow?

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BowItemMixin.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+import net.minecraft.item.BowItem;
+import net.minecraft.item.ItemStack;
+import net.minecraft.world.World;
+
+@Mixin(BowItem.class)
+public class BowItemMixin {
+
+	@Inject(method = "onStoppedUsing(Lnet/minecraft/item/ItemStack;Lnet/minecraft/world/World;Lnet/minecraft/entity/LivingEntity;I)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/world/World;spawnEntity(Lnet/minecraft/entity/Entity;)Z"))
+	public void onStoppedUsing(ItemStack stack, World world, LivingEntity user, int remainingUseTicks,
+			CallbackInfo info) {
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/BowItemMixin.java]: push onto previous line

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/CrossbowItemMixin.java [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+import net.minecraft.item.ItemStack;
+import net.minecraft.world.World;
+
+@Mixin(CrossbowItem.class)
+public class CrossbowItemMixin {
+
+	@Inject(method = "createArrow", at = @At(value = "RETURN"), locals = LocalCapture.CAPTURE_FAILSOFT)
+	private static void createArrow(World world, LivingEntity entity, ItemStack crossbow, ItemStack arrow,
+			CallbackInfoReturnable<PersistentProjectileEntity> cir, ArrowItem arrowItem,
+			PersistentProjectileEntity persistentProjectileEntity) {
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/CrossbowItemMixin.java]: THERE IS ONE LINE

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/CrossbowItemMixin.java [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+
+		if (!(crossbow.getItem() instanceof CrossbowItem)) {
+			return;
+		}
+		((FabricCrossbowHooks) crossbow.getItem()).createArrow(arrowItem, persistentProjectileEntity);
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/CrossbowItemMixin.java]: place a space here after closing the if block, checkstyle violation here.

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/CrossbowItemMixin.java [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+	}
+
+	@Inject(method = "getSpeed", at = @At(value = "HEAD"))
+	private static void tickMovement(ItemStack itemStack, CallbackInfoReturnable<Float> cir) {
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/CrossbowItemMixin.java]: `tickMovement`? This I assume gets the speed at which the crossbow is charged at.

**Comment by OroArmor** [comment]
aight so the builds failed but i changed nothing so...

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Bow.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import net.minecraft.entity.mob.AbstractSkeletonEntity;
+import net.minecraft.item.BowItem;
+import net.minecraft.item.RangedWeaponItem;
+
+@Mixin({ AbstractSkeletonEntity.class })
+public class MobEntityCanUseRangedWeapon_Bow {
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Bow.java]: we do not name mixins this way.

should be `AbstractSkeletonEntityMixin`

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Bow.java [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+import org.spongepowered.asm.mixin.Mixin;
+import org.spongepowered.asm.mixin.injection.At;
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import net.minecraft.entity.mob.AbstractSkeletonEntity;
+import net.minecraft.item.BowItem;
+import net.minecraft.item.RangedWeaponItem;
+
+@Mixin({ AbstractSkeletonEntity.class })
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Bow.java]: the `{ }` is not needed here

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Crossbow.java [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+import org.spongepowered.asm.mixin.injection.Inject;
+import org.spongepowered.asm.mixin.injection.callback.CallbackInfoReturnable;
+
+import net.minecraft.entity.mob.PiglinEntity;
+import net.minecraft.entity.mob.PillagerEntity;
+import net.minecraft.item.CrossbowItem;
+import net.minecraft.item.RangedWeaponItem;
+
+@Mixin({ PiglinEntity.class, PillagerEntity.class })
+public class MobEntityCanUseRangedWeapon_Crossbow {
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Crossbow.java]: Needs to end with a `Mixin` suffix.

Also the name is not suitable, maybe `EntityCrossbowUseMixin`

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Bow.java (hunk 7 above)
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Bow.java]: Illusioners can use bows I think, verify if it is needed there

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ProjectileUtilMixin.java [hunk 9]
```diff
@@ -0,0 +1,52 @@
...
+import net.minecraft.item.Item;
+import net.minecraft.item.Items;
+import net.minecraft.util.Hand;
+
+@Mixin(ProjectileUtil.class)
+public class ProjectileUtilMixin {
+
+	@Inject(method = "getHandPossiblyHolding", at = @At(value = "HEAD"), cancellable = true)
+	private static void getHandPossiblyHolding(LivingEntity entity, Item item, CallbackInfoReturnable<Hand> cir) {
+		if (item == Items.CROSSBOW) {
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ProjectileUtilMixin.java]: this check will fail with any custom crossbow; should be an instanceof check.

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ProjectileUtilMixin.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+			boolean inMainHand = entity.getStackInHand(Hand.MAIN_HAND).getItem() instanceof CrossbowItem;
+			boolean inOffHand = entity.getStackInHand(Hand.OFF_HAND).getItem() instanceof CrossbowItem;
+
+			if (inMainHand || inOffHand) {
+				cir.setReturnValue(inMainHand ? Hand.MAIN_HAND : Hand.OFF_HAND);
+			}
+		} else if (item == Items.BOW) {
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ProjectileUtilMixin.java]: this check will fail with any custom bow; should be an instanceof check

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/client/HeldItemRendererMixin.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+import net.minecraft.client.render.item.HeldItemRenderer;
+import net.minecraft.item.BowItem;
+import net.minecraft.item.CrossbowItem;
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.Items;
+
+@Mixin(HeldItemRenderer.class)
+public class HeldItemRendererMixin {
+
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/client/HeldItemRendererMixin.java]: drop this space

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/client/HeldItemRendererMixin.java [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+	@Redirect(method = "renderFirstPersonItem", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;getItem()Lnet/minecraft/item/Item;"))
+	private Item renderFirstPersonItem(ItemStack itemStack) {
+		return itemStack.getItem() instanceof CrossbowItem ? Items.CROSSBOW : itemStack.getItem();
+	}
+
+	@Redirect(method = "renderItem(FLnet/minecraft/client/util/math/MatrixStack;Lnet/minecraft/client/render/VertexConsumerProvider$Immediate;Lnet/minecraft/client/network/ClientPlayerEntity;I)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;getItem()Lnet/minecraft/item/Item;"))
+	private Item renderItem(ItemStack itemStack) {
+		return itemStack.getItem() instanceof CrossbowItem ? Items.CROSSBOW
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/client/HeldItemRendererMixin.java]: Really should be an if statement rather than two tenary operators for readability:

```suggestion
		if (itemStack.getItem() instanceof CrossbowItem) {
			return Items.CROSSBOW; // Return crossbow for rendering
		}

		if (itemStack.getItem() instanceof BowItem) {
			return Items.BOW; // Return bow for rendering
		}

		return itemStack.getItem();
```

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/client/HeldItemRendererMixin.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+import net.minecraft.item.Item;
+import net.minecraft.item.ItemStack;
+import net.minecraft.item.Items;
+
+@Mixin(HeldItemRenderer.class)
+public class HeldItemRendererMixin {
+
+	@Redirect(method = "renderFirstPersonItem", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;getItem()Lnet/minecraft/item/Item;"))
+	private Item renderFirstPersonItem(ItemStack itemStack) {
+		return itemStack.getItem() instanceof CrossbowItem ? Items.CROSSBOW : itemStack.getItem();
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/client/HeldItemRendererMixin.java]:
```suggestion
		if (itemstack.getItem() instanceof CrossbowItem) {
			return Items.CROSSBOW; // Return true to invoke crossbow rendering path 
		}

		return itemStack.getItem();
```

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/client/HeldItemRendererMixin.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+	}
+
+	@Redirect(method = "renderItem(FLnet/minecraft/client/util/math/MatrixStack;Lnet/minecraft/client/render/VertexConsumerProvider$Immediate;Lnet/minecraft/client/network/ClientPlayerEntity;I)V", at = @At(value = "INVOKE", target = "Lnet/minecraft/item/ItemStack;getItem()Lnet/minecraft/item/Item;"))
+	private Item renderItem(ItemStack itemStack) {
+		return itemStack.getItem() instanceof CrossbowItem ? Items.CROSSBOW
+				: itemStack.getItem() instanceof BowItem ? Items.BOW : itemStack.getItem();
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/client/HeldItemRendererMixin.java]:
```suggestion
```

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/client/ItemStackMixin.java [hunk 15]
```diff
@@ -23,17 +23,18 @@
...
+import net.fabricmc.fabric.api.client.item.v1.ItemTooltipCallback;
 import net.minecraft.client.item.TooltipContext;
 import net.minecraft.entity.player.PlayerEntity;
 import net.minecraft.item.ItemStack;
 import net.minecraft.text.Text;
 
-import net.fabricmc.fabric.api.client.item.v1.ItemTooltipCallback;
-
 @Mixin(ItemStack.class)
 public abstract class ItemStackMixin {
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/client/ItemStackMixin.java]: unnecessary noise for this pull request, revert changes to this class

File: fabric-item-api-v1/src/main/resources/fabric-item-api-v1.mixins.json [hunk 16]
```diff
@@ -4,10 +4,19 @@
   "compatibilityLevel": "JAVA_8",
   "mixins": [
     "ItemMixin",
-    "MobEntityMixin"
+    "MobEntityMixin",
+	"CrossbowItemMixin",
```
Comment (i509VCB) [inline fabric-item-api-v1/src/main/resources/fabric-item-api-v1.mixins.json]: Keep consistent spacing with current file I guess. We need to look into enforcing editor config here and or checkstyle for json files.

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/FabricBowTests.java [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+import net.minecraft.item.ItemGroup;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.Identifier;
+import net.minecraft.util.registry.Registry;
+
+public class FabricBowTests implements ModInitializer {
+	@Override
+	public void onInitialize() {
+		// Registers an item with a custom equipment slot.
+		Item testItem = new TestBow(new FabricItemSettings().group(ItemGroup.COMBAT));
```
Comment (i509VCB) [inline fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/FabricBowTests.java]: You shouldn't need `FabricItemSettings` for setting the item group?

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/FabricBowTests.java [hunk 18]
```diff
@@ -0,0 +1,50 @@
...
+
+	public static class TestBow extends BowItem implements FabricBowHooks {
+
+		public TestBow(Settings settings) {
+			super(settings);
+		}
+
+		@Override
+		public void onBowRelease(ItemStack arrowStack, LivingEntity user, int remainingUseTicks) {
+
```
Comment (i509VCB) [inline fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/FabricBowTests.java]: drop this space, should be staggered brackets.

File: fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/client/TooltipTests.java [hunk 19]
```diff
@@ -16,13 +16,12 @@
 
 package net.fabricmc.fabric.test.item.client;
 
-import net.minecraft.text.LiteralText;
```
Comment (i509VCB) [inline fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/client/TooltipTests.java]: Unnecessary formatting noise, import changes need reverting.

**Review by i509VCB — CHANGES_REQUESTED** [review]
There are a bunch of style issues that need resolving before any further review.

**Comment by OroArmor** [comment]
@i509VCB I fixed all the issues you mentioned.

(Edit Github didnt show all of them to me)

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Bow.java (hunk 7 above)
Comment (OroArmor) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Bow.java]: This is for when new bow mobs are added

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ProjectileUtilMixin.java (hunk 9 above)
Comment (OroArmor) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/ProjectileUtilMixin.java]: when code is called, it is always passing in Items.CROSSBOW or Items.BOW. thats what im checking for. the vanilla code will work for custom items

File: fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Bow.java (hunk 6 above)
Comment (OroArmor) [reply fabric-item-api-v1/src/main/java/net/fabricmc/fabric/mixin/item/MobEntityCanUseRangedWeapon_Bow.java]: EntityBowUseMixin

**Comment by OroArmor** [comment]
All right this is being (maybe) turned into a new module that can be merged in to help remove the hardcoded nature of the minecraft code

**Comment by OroArmor** [comment]
Closing to move to new api
